doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -11,92 +11,79 @@ These tactics use the equality :g:`eq:forall A:Type, A->A->Prop` defined in
+   Some of the variables :g:`x`\ :sub:`i` are solved by unification,
+   and some of the types :n:`A__1, ..., A__n` may become new
+   subgoals.
 
-      Uses the equality :n:`@term`:sub:`1` :n:`= @term` :sub:`2` from right to left
+   :n:`{+, @oriented_rewriter }`
+     The :n:`@oriented_rewriter`\s are applied sequentially
+     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
+     the tactic fails.
 
-   .. tacv:: rewrite @term in @goal_occurrences
+   :n:`{? {| -> | <- } }`
+     For `->` (the default), :n:`term__1` is rewritten
+     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term_1`.
 
-      Analogous to :n:`rewrite @term` but rewriting is done following
-      the clause :token:`goal_occurrences`. For instance:
+   :n:`@natural`
+     If followed by :n:`?`, gives the maximum number of rewrites that will be performed.
+     Otherwise, it specifies the exact number of rewrites to perform (whether followed by
+     `!` or not).  If the given number of rewrites can't be done, the tactic fails.
 
-      + :n:`rewrite H in H'` will rewrite `H` in the hypothesis
-        ``H'`` instead of the current goal.
-      + :n:`rewrite H in H' at 1, H'' at - 2 |- *` means
-        :n:`rewrite H; rewrite H in H' at 1; rewrite H in H'' at - 2.`
-        In particular a failure will happen if any of these three simpler tactics
-        fails.
-      + :n:`rewrite H in * |-` will do :n:`rewrite H in H'` for all hypotheses
-        :g:`H'` different from :g:`H`.
-        A success will happen as soon as at least one of these simpler tactics succeeds.
-      + :n:`rewrite H in *` is a combination of :n:`rewrite H` and :n:`rewrite H in * |-`
-        that succeeds if at least one of these two tactics succeeds.
+   :n:`?`
+     Performs the rewrite as many times as possible (possibly zero times). This form never fails.
 
-      Orientation :g:`->` or :g:`<-` can be inserted before the :token:`term` to rewrite.
+   :n:`!`
+     Performs the rewrite as many times as possible and at least once.  Otherwise this
+     form fails.
 
-   .. tacv:: rewrite @term at @occurrences
+   :n:`@occurrences`
+     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
+     any of them can be rewritten.  If not specified, only the first occurrence
+     in the conclusion is replaced.
 
-      Rewrite only the given :token:`occurrences` of :token:`term`. Occurrences are
-      specified from left to right as for pattern (:tacn:`pattern`). The rewrite is
-      always performed using setoid rewriting, even for Leibniz’s equality, so one
-      has to ``Import Setoid`` to use this variant.
+     If :n:`at @occs_nums` is specified, rewriting is always done with setoid
+     rewriting, even for Leibniz’s equality, so you must
+     ``Import Setoid`` to use this form.
 
-   .. tacv:: rewrite @term by @tactic
+     .. todo PR above is correct?  rewrite ... at 1 2 3 works even if only 1 is rewritten?
```

Comment:
Apparently, yes:

```coq
Require Setoid.
Goal 1 = 0 -> 0 = 1.
intro H.
rewrite H at 1 2 3.
```

But shouldn't it be considered as a bug to fix (that it does not fail even if the occurrence doesn't exist)?

Reply:
Yup.  https://github.com/coq/coq/issues/13566